Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx

```diff
@@ -230,14 +266,288 @@ ContourExtractor2DImageFilter<TInputImage>::GenerateData()
+      const InputPixelType   differentLabel = label + 1;
+      std::fill(stripeRange.begin(), stripeRange.end(), differentLabel);
+    }
+    const IndexType       shrunkIndex{ min[0] - 1, min[1] - 1 };
+    const SizeType        shrunkSize{ static_cast<SizeValueType>(max[0] - min[0]) + 2,
+                               static_cast<SizeValueType>(max[1] - min[1]) + 2 };
+    const InputRegionType shrunkRegion{ shrunkIndex, shrunkSize };
+
+    // A 1-pixel radius sets up a neighborhood with the following indices:
+    // 0 1 2
+    // 3 4 5
+    // 6 7 8
+    // We are interested only in the square of 4,5,7,8 which is the 2x2 square
+    // with the center pixel at the top-left. So we only activate the
+    // coresponding offsets, and only query pixels 4, 5, 7, and 8 with the
+    // iterator's GetPixel method.
+    SquareIterator it{ radius, largerImage, shrunkRegion };
+    it.ActivateOffset(none);
+    it.ActivateOffset(right);
+    it.ActivateOffset(down);
+    it.ActivateOffset(diag);
+
+    for (it.GoToBegin(); !it.IsAtEnd(); ++it)
+    {
+      // There are sixteen different possible square types, diagramed below.
+      // A + indicates that the vertex has the current label, and a -
+      // indicates that the vertex is does not have the current label.
+      // The vertices of each square are here numbered:
+      // 01
+      // 23
+      // and treated as a binary value with the bits in that order. Thus each
+      // square can be so numbered:
+      //  0--   1+-   2-+   3++   4--   5+-   6-+   7++
+      //   --    --    --    --    +-    +-    +-    +-
+      //
+      //  8--   9+-  10-+  11++  12--  13+-  14-+  15++
+      //   -+    -+    -+    -+    ++    ++    ++    ++
+      //
+      // The position of the line segment that cuts through (or doesn't, in case
+      // 0 and 15) each square is clear, except in cases  6 and 9. In this case,
+      // where the segments are placed is determined by
+      // m_VertexConnectHighPixels. If m_VertexConnectHighPixels is false, then
+      // lines like are drawn through square 6, and lines like are drawn through
+      // square 9. Otherwise, the situation is reversed.
+      // Finally, recall that we draw the lines so that (moving from tail to
+      // head) the non-labeled pixels are on the left of the line. So, for
+      // example, case 1 entails a line slanting from the middle of the top of
+      // the square to the middle of the left side of the square.
+
+      // (1) Determine what number square we are currently inspecting. Remember
+      // that as far as the neighborhood iterator is concerned, our square
+      // 01    is numbered as    45
+      // 23                      78
+
+      const bool           v0{ it.GetPixel(4) == label };
+      const bool           v1{ it.GetPixel(5) == label };
+      const bool           v2{ it.GetPixel(7) == label };
+      const bool           v3{ it.GetPixel(8) == label };
```

Comment:
FYI: The initialization of `v0` to `v3` could look as follows when using ITK5 `ShapedImageNeighborhoodRange` (with `offsets = { none, right, down, diag }`):

      const ShapedImageNeighborhoodRange<const InputImageType> neighborhoodRange(*largerImage, index, offsets);

      const bool           v0{ neighborhoodRange[0] == label };
      const bool           v1{ neighborhoodRange[1] == label };
      const bool           v2{ neighborhoodRange[2] == label };
      const bool           v3{ neighborhoodRange[3] == label };

Reply:
This is now Issue #2348.